File: 4th/homework_4/id_45/70_climbing_stairs.cpp
//Solution 1
class Solution {
public:
    int climbStairs(int n) {
        vector<int> result(n + 1, -1);
        result[0] = result[1] = 1;
        climbWays(result, n);
        return result[n];
    }
    int climbWays(vector<int>& result, int level) {
        if (result[level] != -1)
            return result[level];
        return result[level] = climbWays(result, level - 1) + climbWays(result, level - 2);
    }
};

//Solution 2: 
class Solution {
public:
    int climbStairs(int n) {
        vector<int> result(n + 1, -1);
        result[0] = result[1] = 1;
        for (int i = 2; i < n + 1; ++i) {
            result[i] = result[i - 1] + result[i - 2];
        }
        return result[n];
    }
};

File: 4th/homework_4/id_69/70-ClimbingStairs.py
# https://leetcode.com/problems/climbing-stairs/description/
# You are climbing a stair case. It takes n steps to reach to the top.

# Each time you can either climb 1 or 2 steps.
# In how many distinct ways can you climb to the top?

# Note: Given n will be a positive integer.
# dp[n] = dp[n-1] + dp[n-2]

class Solution(object):
    def climbStairs(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n == 0: return 0
        dp = [0 for _ in range(n+1)]
        dp[0] = dp[1] = 1
        for i in range(2, n + 1):
            dp[i] = dp[i-1] + dp[i-2]
        return dp[n]

if __name__ == '__main__':
    sol = Solution()
    for i in range(10):
        print(sol.climbStairs(i))
